### DISCUSSION TITLE
A more visual graph view

### DISCUSSION THREAD

**Discussion by gillchristian** [srid/neuron#99] [Ideas]
It would be really neat to haven an alternative view of the Zettelkasten graph that could be navigated.

[Example](https://learn-anything.xyz/computer-science/data-structures) 👇 from [github.com/learn-anything/learn-anything](https://github.com/learn-anything/learn-anything)

<img width="861" alt="Screenshot 2020-04-13 at 18 57 58" src="https://user-images.githubusercontent.com/8309423/79140719-b3125900-7db8-11ea-9893-2976d30deb2d.png">

NOTE: I'm up to work on this. Probably better to do after #98

**Comment by srid** [comment]
> Probably better to do after #98

Agreed. I'd feel a little uncomfortable if the amount of _JavaScript_ on this repo begins to grow :stuck_out_tongue:

**Comment by srid** [comment]
On that note, if you wish to take a stab at #98 -- so as to port over `./src-script/search.js` to PureScript -- feel free to go ahead (but do let me know what your approach would be)!

UPDATE (Oct 9): Let's use GHCJS (reflex), instead of PureScript.

**Comment by gillchristian** [comment]
I'm not familiar with PS ecosystem nor with nix so it'll take some time, but when/if I find something I'll post on #98

**Comment by srid** [comment]
Some improvements in #195

**Comment by jaygray0919** [comment]
Might it make sense to export to a graph generator? For example, the classic RDF graph generator is [Rhizomik](http://rhizomik.net/html/redefer/rdf2svg-form/)  which uses [Graphviz](https://www.graphviz.org/).  An example of the dot notation and rendering is [here](http://arborjs.org/halfviz/#/date-with-destiny) and [here](https://graphviz.christine.website/#digraph%20G%20%7B%0A%0A%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20style%3Dfilled%3B%0A%20%20%20%20color%3Dlightgrey%3B%0A%20%20%20%20node%20%5Bstyle%3Dfilled%2Ccolor%3Dwhite%5D%3B%0A%20%20%20%20a0%20-%3E%20a1%20-%3E%20a2%20-%3E%20a3%3B%0A%20%20%20%20label%20%3D%20%22process%20%231%22%3B%0A%20%20%7D%0A%0A%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20b0%20-%3E%20b1%20-%3E%20b2%20-%3E%20b3%3B%0A%20%20%20%20label%20%3D%20%22process%20%232%22%3B%0A%20%20%20%20color%3Dblue%0A%20%20%7D%0A%20%20start%20-%3E%20a0%3B%0A%20%20start%20-%3E%20b0%3B%0A%20%20a1%20-%3E%20b3%3B%0A%20%20b2%20-%3E%20a3%3B%0A%20%20a3%20-%3E%20a0%3B%0A%20%20a3%20-%3E%20end%3B%0A%20%20b3%20-%3E%20end%3B%0A%0A%20%20start%20%5Bshape%3DMdiamond%5D%3B%0A%20%20end%20%5Bshape%3DMsquare%5D%3B%0A%7D).
[Mermaid](https://mermaid-js.github.io/mermaid/#/) is used on GitHub.  Here is an online [example](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbiAgQVtDaHJpc3RtYXNdIC0tPnxHZXQgbW9uZXl8IEIoR28gc2hvcHBpbmcpXG4gIEIgLS0-IEN7TGV0IG1lIHRoaW5rfVxuICBDIC0tPnxPbmV8IERbTGFwdG9wXVxuICBDIC0tPnxUd298IEVbaVBob25lXVxuICBDIC0tPnxUaHJlZXwgRltmYTpmYS1jYXIgQ2FyXVxuXHRcdCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In19).
A new entry is [diagram.code](https://www.diagram.codes/)

The basic idea:

> export Neuron data as input notation to the preferred graph engine (and template).

Alternatively, like Markdown support for Mermaid on GitHub, implement the graph in a specific zettel, with links to navigate to target zettels.

**Comment by GenrihGrigoryan** [comment]
The perfect ability for making visual graph is to copy it from the obsidian realisation:
![1*bBhfq-ulQybUTkr-YKbuAg](https://user-images.githubusercontent.com/7505178/101984508-e2080400-3c92-11eb-8fa4-f5711dd4293f.png)

**Comment by srid** [comment]
I don't find these graph views very useful for navigation -- they provide more of eye-candy than utility -- however my plan is to use that only as a starting point (our current z-index UX stops being useful for larger zettelkastens), and then refine it further (use folgezettel, 'graph search' narrowing, etc.).  See [Graphic views do not work with a Zettelkasten](https://www.reddit.com/r/Zettelkasten/comments/jogsw0/graphic_views_do_not_work_with_a_zettelkasten/). I'm experimenting with these ideas [here](https://github.com/srid/rememorate), which will eventually replace both the current search and z-index pages.

@aca made this prototype: https://aca.github.io/neuron_zettels_in_3d.html

**Comment by jaygray0919** [comment]
We agree
> I don't find these graph views very useful for navigation
Graph navigation should be `similarTo` menu navigation where the submenu are exposed/visible (i.e. not revealsed on mouseover).

It's certainly important to support complex documents (mermaid; d3.js; three.js, graphviz) on a zettel page.

But those services are implicitly/explicitly designed for a canvas - not a necessarily compact format such as navigation (menu selector).

However, we are concerned about maintainability of neuron instances.
If we understand correctly, the formatting services (mermaid, graphviz, ...) are page specific and manually injected.
In contrast, Sphinx requires services to be defined in a configuration.py document such that the service can be used on any page that is part of the configuration.

MathJax support in Neuron may be an example.
Since MathJax is "configured" the notation can be used on any zettel.

Might it be better (i.e. easier to maintain) if neuron required services to be configured globally, and then used by declaring a service identifier before including processor-specific code (e.g. d3.js or three.js).?

**Comment by srid** [comment]
@jaygray0919 They don't have to be page-specific. You can put the JavaScript code globally in a file named `head.html` (MathJax is included in this way) and they get automatically included in all pages. See https://neuron.zettel.page/custom-head.html

The menu navigation is a good idea.
